File: src/dataset/python/bowling_exercism-ThomasZumsteg/openAI/ChatGPT4_bowling_v4.py
class BowlingGame:
    def __init__(self):
        self._rolls = []
        self._frame = 0
        self._extra_rolls = 0

    def roll(self, pins):
        if self._game_over():
            raise IndexError("Game Over")
        if not (0 <= pins <= 10):
            raise ValueError(f"Cannot score {pins}")
        if self._frame % 2 == 1 and self._rolls[-1] + pins > 10 and self._frame < 18:
            raise ValueError(f"Cannot score {self._rolls[-1]} and {pins}")

        self._rolls.append(pins)
        self._update_state(pins)

    def score(self):
        if not self._game_over():
            raise IndexError("Game in progress")

        total, index = 0, 0
        for frame in range(10):
            if self._rolls[index] == 10:  # Strike
                total += 10 + self._rolls[index + 1] + self._rolls[index + 2]
                index += 1
            elif sum(self._rolls[index:index + 2]) == 10:  # Spare
                total += 10 + self._rolls[index + 2]
                index += 2
            else:  # Open frame
                total += sum(self._rolls[index:index + 2])
                index += 2
        return total

    def _update_state(self, pins):
        if self._frame < 18:  # Regular frames
            self._frame += 2 if pins == 10 and self._frame % 2 == 0 else 1
        elif self._frame == 18:  # 10th frame
            self._extra_rolls = 2 if pins == 10 else 1 if sum(self._rolls[-2:]) == 10 else 0
            self._frame += 1
        elif self._extra_rolls > 0:  # Extra rolls
            self._extra_rolls -= 1

    def _game_over(self):
        return self._frame >= 20 and self._extra_rolls == 0

File: src/dataset/javascript/bowling_exercism-ThomasZumsteg/openAI/ChatGPT4_bowling_v4.js
class Bowling {
    constructor() {
        this.rolls = [];
        this.frame = 0;
        this.extra = 0;
        this.second = false;
    }

    gameOver() {
        return this.frame >= 10 && this.extra <= 0;
    }

    roll(pins) {
        if (this.gameOver()) throw new Error('Cannot roll after game is over');
        if (pins < 0) throw new Error('Negative roll is invalid');
        if (pins > 10 || (this.second && this.rolls[this.rolls.length - 1] + pins > 10)) {
            throw new Error('Pin count exceeds pins on the lane');
        }

        if (this.frame === 9) {
            if (!this.second && pins === 10) {
                this.extra = 2;
            } else if (this.second && this.rolls[this.rolls.length - 1] + pins === 10) {
                this.extra = 1;
            }
        } else if (this.frame > 9) {
            this.extra--;
        }

        if (pins === 10) {
            this.frame++;
            this.second = false;
        } else {
            if (this.second) this.frame++;
            this.second = !this.second;
        }
        this.rolls.push(pins);
    }

    score() {
        if (!this.gameOver()) throw new Error('Score cannot be taken until the end of the game');

        let result = 0, i = 0;
        for (let frame = 0; frame < 10; frame++) {
            const frameScore = this.rolls[i] + (this.rolls[i + 1] || 0);
            result += frameScore;

            if (this.rolls[i] === 10) {
                result += (this.rolls[i + 1] || 0) + (this.rolls[i + 2] || 0);
                i++;
            } else if (frameScore === 10) {
                result += this.rolls[i + 2] || 0;
                i += 2;
            } else {
                i += 2;
            }
        }
        return result;
    }
}

module.exports = Bowling;